File: rust/introductory/palindrome_reorder.rs
use std::io;

fn main() {
    let mut buffer = String::new();
    io::stdin().read_line(&mut buffer);
    let buffer = buffer.trim();

    let mut cnt = [0; 30];

    for c in buffer.as_bytes() {
        cnt[(c - ('A' as u8)) as usize] += 1;
    }

    let mut index: i32 = -1;

    for i in 0..30 {
        if index != -1 && cnt[i as usize] % 2 == 1 {
            println!("NO SOLUTION");
            return;
        } else if cnt[i as usize] % 2 == 1 {
            index = i as i32;
        }
    }

    for i in 0..30 {
        if i == index {
            continue;
        }

        for j in 0..(cnt[i as usize] / 2) {
            print!("{}", (i as u8 + 'A' as u8) as char);
        }
    }

    if index != -1 {
        for j in 0..(cnt[index as usize]) {
            print!("{}", (index as u8 + 'A' as u8) as char);
        }
    }

    for i in (0..30).rev() {
        if i == index {
            continue;
        }

        for j in 0..(cnt[i as usize] / 2) {
            print!("{}", (i as u8 + 'A' as u8) as char);
        }
    }
}

File: cpp/introductory/palindrome_reorder.cpp
#include <bits/stdc++.h>

using namespace std;

#define endl '\n'
#define f1 first
#define s2 second
#define pb push_back
#define mp make_pair
#define int long long
#define fri(a) freopen(a,"r",stdin);
#define fro(a) freopen(a,"w",stdout);
const int MOD = 1e9 + 7;
const int N = 30;

string s;
int cnt[N];

int32_t main() {

    //fri("in.txt");
    //fro("out.txt");

    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> s;

    for (auto& c : s) {
        cnt[c - 'A']++;
    }


    int index = -1;

    for (int i = 0; i < N; i++) {
        if (index != -1 && (cnt[i] & 1)) {
            cout << "NO SOLUTION" << endl;
            return 0;
        }
        else if (cnt[i] & 1) {
            index = i;
        }
    }


    for (int i = 0; i < N; i++) {
        if (index == i) {
            continue;
        }
        for (int j = 0; j < cnt[i] / 2; j++) {
            cout << (char)('A' + i);
        }
    }

    if (index != -1) {
        for (int j = 0; j < cnt[index]; j++) {
            cout << (char)('A' + index);
        }
    }

    for (int i = N; i >= 0; i--) {
        if (index == i) {
            continue;
        }
        for (int j = 0; j < cnt[i] / 2; j++) {
            cout << (char)('A' + i);
        }
    }




    return 0;
}
